darling/src/libsystem_kernel/emulation/src/xnu_syscall/bsd/impl/misc/clonefile.c

```diff
@@ -1,13 +1,224 @@
 #include <darling/emulation/xnu_syscall/bsd/impl/misc/clonefile.h>
 
+#include <sys/errno.h>
+#include <sys/stat.h>
+#include <sys/time.h>
 #include <signal.h>
+#include <string.h>
+#include <time.h>
 
-#include <darling/emulation/conversion/duct_errno.h>
+#include <darling/emulation/common/base.h>
+#include <darling/emulation/conversion/common_at.h>
+#include <darling/emulation/conversion/errno.h>
+#include <darling/emulation/conversion/fcntl/open.h>
+#include <darling/emulation/linux_premigration/linux-syscalls/linux.h>
+#include <darling/emulation/linux_premigration/vchroot_expand.h>
+#include <darling/emulation/other/mach/lkm.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/fcntl/open.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/fcntl/openat.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/stat/mkdirat.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/unistd/close.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/unistd/fchmod.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/unistd/fchown.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/unistd/unlink.h>
+
+#ifndef MAX_COPY_CHUNK
+#define MAX_COPY_CHUNK ((size_t)0x7ffff000)
+#endif
 
 long sys_clonefileat(int src_fd, const char* src_path, int dest_fd, const char* dest_path, uint32_t flags) {
 	return -ENOSYS;
 }
 
 long sys_fclonefileat(int src_fd, int dest_fd, const char* dest_path, uint32_t flags) {
-	return -ENOSYS;
+    struct stat st;
+
+    if (fstat(src_fd, &st) < 0) {
+        return -errno;
+    }
+
+    if (!S_ISREG(st.st_mode)) {
+        return -EOPNOTSUPP;
+    }
+
+    if (flags & ~CLONE_NOFOLLOW) {
+        return -EINVAL;
+    }
+
+    int open_flags = BSD_O_CREAT | BSD_O_WRONLY | BSD_O_EXCL;
+
+    if (flags & CLONE_NOFOLLOW) {
+        open_flags |= BSD_O_NOFOLLOW;
+    }
```

Comment:
```
     CLONE_NOFOLLOW         Don't follow the src file if it is a symbolic
                            link (applicable only if the source is not a
                            directory).  The symbolic link is itself cloned
                            if src names a symbolic link.
```

From my testing, it doesn't seem like `CLONE_NOFOLLOW` has any affect when used in `fclonefileat`. But even if it did, your implementation affects the destination, not the source.